### PR TITLE
Fix bundle creation

### PR DESCRIPTION
**Pull request by axifive**
Fix for added bundle download (#14538). `git bundle create` can only accept a reference:

> 2021/09/17 15:37:16 ...archiver/archiver.go:237:func1() [E] Archive &{1 master bundle 8c3a46a6fadf07e24199cf0fbf0fcb973cd5efb8} faild: unable to write archive: exit status 128 - fatal: Refusing to create empty bundle.

### CODE REVIEW

File: services/archiver/archiver.go [hunk 1]
```diff
@@ -135,9 +135,11 @@ func doArchive(r *ArchiveRequest) (*models.RepoArchiver, error) {
 	if err == nil {
 		if archiver.Status == models.RepoArchiverGenerating {
 			archiver.Status = models.RepoArchiverReady
-			return archiver, models.UpdateRepoArchiverStatus(ctx, archiver)
+			if err = models.UpdateRepoArchiverStatus(ctx, archiver); err != nil {
+				return nil, err
+			}
 		}
-		return archiver, nil
+		return archiver, commiter.Commit()
```
Comment (axifive) [inline services/archiver/archiver.go]: Need when archive already exist in storage, but for some reason the record in db not founded(so it created within tx session and then all changes rollbacked here).

**Comment by zeripath** [comment]
Instead of creating a branch in the real repo this should create a temporary repo with GIT_OBJECT_DIRECTORY set to the main repo and then you create your own branch as appropriate, then the whole temporary repo can be deleted at the end without having to worry about issues like what if a panic is called earlier.

We must not go around creating branches in real users repos. They will appear on the UI - even for a short amount of time - but the current code risks the temporary branch not being deleted at the end if there is an error.

**Comment by jolheiser** [comment]
Another alternative, maybe we just don't offer bundle downloads for a commit? (If there's no simple way to provide them, that is)  
I still personally think the majority of use-cases for bundles are the entire main branch/repo.

**Comment by axifive** [comment]
@jolheiser the way suggested by @zeripath is quite simple

**Comment by 6543** [comment]
Why is it kind/breaking?

**Comment by axifive** [comment]
@6543 add multiple by mistake

File: services/archiver/archiver.go [hunk 2]
```diff
@@ -136,9 +136,11 @@ func doArchive(r *ArchiveRequest) (*models.RepoArchiver, error) {
...
```
Comment (gwymor) [inline services/archiver/archiver.go]: Typo breaking build:

```suggestion
		return archiver, committer.Commit()
```

Comment (axifive) [reply services/archiver/archiver.go]: Thanks, not typo, it just has been renamed in main branch

File: modules/git/repo.go [hunk 3]
```diff
@@ -425,14 +425,20 @@ func (repo *Repository) CreateBundle(ctx context.Context, commit string, out io.
...
-	_, err = NewCommandContext(ctx, args...).RunInDir(repo.Path)
+
+	branchName := "main"
+	_, err = NewCommandContext(ctx, "branch", branchName, commit).RunInDirWithEnv(tmp, env)
+	if err != nil {
+		return err
+	}
+
+	tmpFile := filepath.Join(tmp, "bundle")
+	_, err = NewCommandContext(ctx, "bundle", "create", tmpFile, branchName).RunInDirWithEnv(tmp, env)
```
Comment (gwymor) [inline modules/git/repo.go]: Hardcoding `main` here is confusing. If I have `main` and `dev` branches and download a bundle of `dev`, it'll show the branch name `main` with the contents of `dev`.

Ideally the bundle would contain two references, `HEAD` and the name of the branch downloaded if we know that (pointing to the same commit). The `HEAD` is a convenience so people cloning from the bundle don't have to specify a branch.

That would require changing the repo archiver machinery to pass a desired branch name into this function, so for an initial fix I think we should just make a bundle with a `HEAD`.

```suggestion
	_, err = NewCommandContext(ctx, "reset", "--soft", commit).RunInDirWithEnv(tmp, env)
	if err != nil {
		return err
	}

	tmpFile := filepath.Join(tmp, "bundle")
	_, err = NewCommandContext(ctx, "bundle", "create", tmpFile, "HEAD").RunInDirWithEnv(tmp, env)
```

Comment (axifive) [reply modules/git/repo.go]: `git clone` from such a bundle will lead to repo with detached HEAD. Probably better set "bundle" reference

> Hardcoding main here is confusing. If I have main and dev branches and download a bundle of dev, it'll show the branch name main with the contents of dev

yes, **reponame-dev.bundle** with `main` branch inside it.

> That would require changing the repo archiver machinery to pass a desired branch name into this function, so for an initial fix I think we should just make a bundle with a HEAD.

It's not a problem to send the name of the branch, the problem if that the different branches include the same commit, but the commit field is unique in the database. So if archive already exists it returns with random branch

Comment (gwymor) [reply modules/git/repo.go]: I see, okay.

Perhaps in the future it would be possible to generate the ref headers on the fly, while reusing the existing bundle file. [The bundle format](https://git-scm.com/docs/bundle-format) is not too complicated (essentially a few header lines concatenated with a Git packfile), so we may be able to:

1. Read the headers of the existing bundle
2. Use that info to construct a new header with the correct ref and send that to the client
3. Seek to the end of the header section in the bundle and send that to the client

Without impacting download time/storage.

(Of course, I don't mean to suggest that you need to do any of this.)

Comment (axifive) [reply modules/git/repo.go]: Better to create an issue with this proposal, so it doesn't get lost

**Review by wxiaoguang — APPROVED** [review]

File: services/archiver/archiver.go [hunk 4]
```diff
@@ -136,9 +136,11 @@ func doArchive(r *ArchiveRequest) (*models.RepoArchiver, error) {
 	if err == nil {
 		if archiver.Status == models.RepoArchiverGenerating {
 			archiver.Status = models.RepoArchiverReady
-			return archiver, models.UpdateRepoArchiverStatus(ctx, archiver)
+			if err = models.UpdateRepoArchiverStatus(ctx, archiver); err != nil {
```
Comment (6543) [inline services/archiver/archiver.go:139]: this is unrelated and should be backported

Comment (axifive) [reply services/archiver/archiver.go:139]: ready #17151

**Review by 6543 — APPROVED** [review]
